src/worker.cc

```diff
@@ -173,27 +165,54 @@ rapidjson::Document from_string(const std::string& json, const valhalla_exceptio
   return d;
 }
 
-void add_date_to_locations(Options& options,
-                           google::protobuf::RepeatedPtrField<valhalla::Location>& locations) {
-  // otherwise we do what the person was asking for
+bool add_date_to_locations(Options& options,
+                           google::protobuf::RepeatedPtrField<valhalla::Location>& locations,
+                           const std::string& node) {
   if (options.has_date_time_case() && !locations.empty()) {
-    switch (options.date_time_type()) {
-      case Options::current:
-        locations.Mutable(0)->set_date_time("current");
-        break;
-      case Options::depart_at:
-        locations.Mutable(0)->set_date_time(options.date_time());
-        break;
-      case Options::arrive_by:
-        locations.Mutable(locations.size() - 1)->set_date_time(options.date_time());
-        break;
-      case Options::invariant:
-        for (auto& loc : locations)
-          loc.set_date_time(options.date_time());
-      default:
-        break;
+    if (options.action() != Options::sources_to_targets) {
+      switch (options.date_time_type()) {
+        case Options::current:
+          locations.Mutable(0)->set_date_time("current");
+          break;
+        case Options::depart_at:
+          locations.Mutable(0)->set_date_time(options.date_time());
+          break;
+        case Options::arrive_by:
+          locations.Mutable(locations.size() - 1)->set_date_time(options.date_time());
+          break;
+        case Options::invariant:
+          for (auto& loc : locations)
+            loc.set_date_time(options.date_time());
+        default:
+          break;
+      }
+    } else {
+      switch (options.date_time_type()) {
+        case Options::current:
+        case Options::depart_at:
+          if (node == "sources") {
+            for (auto& loc : locations) {
+              loc.set_date_time(options.date_time_type() == Options::current ? "current"
+                                                                             : options.date_time());
+            }
+          }
+          break;
```

Comment:
this will eventually also be interesting for isochrones & centroid, not only matrix.